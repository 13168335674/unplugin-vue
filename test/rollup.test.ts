import { resolve } from 'path'
import { describe, it, expect } from 'vitest'
import { rollup } from 'rollup'
import glob from 'fast-glob'
import ViteVue from '@vitejs/plugin-vue'
import * as vueCompiler from 'vue/compiler-sfc'
import Vue from '../src/rollup'
import type { Options } from '../src'

async function getCode(file: string, plugin: any) {
  const bundle = await rollup({
    input: [file],
    external: ['vue'],
    plugins: [plugin],
  })
  const output = await bundle.generate({ format: 'esm' })
  return output.output
    .map((file) => {
      if (file.type === 'chunk') {
        return file.code
      } else {
        return file.fileName
      }
    })
    .join('\n')
}

function createPlugins(opt: Options) {
  const vite = ViteVue({
    reactivityTransform: true,
    compiler: opt.compiler!,
  })
  vite.configResolved!({
    root: opt.root!,
    command: 'build',
    isProduction: true,
    build: {
      sourcemap: false,
    },
  } as any)
  return {
    unplugin: Vue(opt),
    vite,
  }
}

describe('transform', () => {
  describe('fixtures', async () => {
    const root = resolve(__dirname, '..')
    const files = await glob('test/fixtures/*.{vue,js,ts}', {
      cwd: root,
      onlyFiles: true,
    })

    for (const file of files) {
      describe(file.replaceAll('\\', '/'), () => {
        const filepath = resolve(root, file)

        for (const isProduction of [true, false]) {
          it(`isProduction is ${isProduction}`, async () => {
            const { unplugin, vite } = createPlugins({
              root,
              compiler: vueCompiler,
              reactivityTransform: true,
              isProduction: true,
            })

            const viteCode = await getCode(filepath, vite)
            const unpluginCode = await getCode(filepath, unplugin)
            expect(unpluginCode).toMatchSnapshot()

            expect(viteCode).toBe(unpluginCode)
          })
        }
      })
    }
  })
})
