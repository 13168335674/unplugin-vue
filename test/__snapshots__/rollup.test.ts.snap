// Vitest Snapshot v1

exports[`transform > fixtures > test/fixtures/basic.vue 1`] = `
"import { openBlock, createElementBlock, createElementVNode } from 'vue';

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

console.log('Hello world');

const _sfc_main = {};

const _hoisted_1 = /*#__PURE__*/createElementVNode(\\"h1\\", null, \\"Hello World\\", -1 /* HOISTED */);
const _hoisted_2 = [
  _hoisted_1
];

function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return (openBlock(), createElementBlock(\\"div\\", null, _hoisted_2))
}
var basic = /*#__PURE__*/_export_sfc(_sfc_main, [['render',_sfc_render]]);

export { basic as default };
"
`;

exports[`transform > fixtures > test/fixtures/reactivity-transform.ts 1`] = `
"import { ref } from 'vue';

const ref1 = ref('hello');
const ref2 = (ref('hello'));

console.log(ref1.value);
console.log(ref2.value);
console.log((ref1));
console.log((ref2));
"
`;

exports[`transform > fixtures > test/fixtures/reactivity-transform.vue 1`] = `
"import { ref } from 'vue';

const _sfc_main = {
  setup(__props, { expose }) {
  expose();

const ref1 = ref('hello');
const ref2 = (ref('hello'));

console.log(ref1.value);
console.log(ref2.value);
console.log((ref1));
console.log((ref2));

const __returned__ = { ref1, ref2, ref };
Object.defineProperty(__returned__, '__isScriptSetup', { enumerable: false, value: true });
return __returned__
}

};

export { _sfc_main as default };
"
`;

exports[`transform > fixtures > test/fixtures/script-setup.vue 1`] = `
"import { openBlock, createElementBlock, createElementVNode } from 'vue';

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

const _sfc_main = {
  setup(__props, { expose }) {
  expose();

const foo = 'bar';


const __returned__ = { foo };
Object.defineProperty(__returned__, '__isScriptSetup', { enumerable: false, value: true });
return __returned__
}

};

const _hoisted_1 = /*#__PURE__*/createElementVNode(\\"h1\\", null, \\"Hello World\\", -1 /* HOISTED */);
const _hoisted_2 = [
  _hoisted_1
];

function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return (openBlock(), createElementBlock(\\"div\\", null, _hoisted_2))
}
var scriptSetup = /*#__PURE__*/_export_sfc(_sfc_main, [['render',_sfc_render]]);

export { scriptSetup as default };
"
`;

exports[`transform > fixtures > test/fixtures/script-setup-ts.vue 1`] = `
"import { defineComponent, openBlock, createElementBlock, createElementVNode } from 'vue';

var _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};

const _sfc_main = /* @__PURE__ */ defineComponent({
  props: {
    foo: { type: String, required: true, default: \\"123\\" },
    bar: { type: Number, required: true, default: 123 },
    baz: { type: Boolean, required: true, default: true }
  },
  setup(__props, { expose }) {
    const str = \\"bar\\";
    expose({
      str
    });
    const __returned__ = { str };
    Object.defineProperty(__returned__, \\"__isScriptSetup\\", { enumerable: false, value: true });
    return __returned__;
  }
});
const _hoisted_1 = /* @__PURE__ */ createElementVNode(\\"h1\\", null, \\"Hello World\\", -1);
const _hoisted_2 = [
  _hoisted_1
];
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock(\\"div\\", null, _hoisted_2);
}
var scriptSetupTs = /* @__PURE__ */ _export_sfc(_sfc_main, [[\\"render\\", _sfc_render]]);

export { scriptSetupTs as default };
"
`;

exports[`transform > fixtures > test/fixtures/typescript.vue 1`] = `
"const _sfc_main = {};

export { _sfc_main as default };
"
`;
